src/wallet/wallet2.h

```diff
@@ -1080,6 +1090,24 @@ namespace tools
       src.real_out_tx_key = get_tx_pub_key_from_extra(td.m_tx);
       src.real_output = interted_it - src.outputs.begin();
       src.real_output_in_tx_index = td.m_internal_output_index;
+      {
+        // read unencrypted payment id needed for onetime addressing
+        std::vector<tx_extra_field> tx_extra_fields;
+        tx_extra_nonce extra_nonce;
+        src.unencrypted_payment_id = null_hash;
+        if (parse_tx_extra(td.m_tx.extra, tx_extra_fields) && find_tx_extra_field_by_type(tx_extra_fields, extra_nonce))
+        {
+          if (!get_payment_id_from_tx_extra_nonce(extra_nonce.nonce, src.unencrypted_payment_id))
+          {
+            crypto::hash8 payment_id8;
+            if (get_encrypted_payment_id_from_tx_extra_nonce(extra_nonce.nonce, payment_id8))
+            {
+              memcpy(src.unencrypted_payment_id.data, payment_id8.data, 8);
+              memset(src.unencrypted_payment_id.data + 8, 0, 24);
+            }
+          }
+        }
+      }
```

Comment:
The repetitive code is moved to `set_unencrypted_payment_id_from_tx_extra(const vector<uint8_t>&, tx_source_entry&)` in `cryptonote_format_util`.